### PR TITLE
Add `clip_gradients` hook in `LightningModule

### LINKED ISSUE #6346: [RFC] Gradient clipping hooks in the LightningModule
## 🚀 Feature

Add clipping hooks to the LightningModule

### Motivation

It's currently very difficult to change the clipping logic

### Pitch

```python
class LightningModule:
    def on_before_clip_gradients():
        ...

    def clip_gradients(self, optimizer):
        ...
    
    def on_after_clip_gradients():
        ...
```

The default implementation would be the same as we currently provide, where the trainer clipping parameters are used.

Maybe those would be deprecated in favor of LightningModule properties.

```python
class LightningOptimizer
    def step(closure=closure)
        if closure is None:
            closure = do_nothing_closure
        def wrapper_closure()
            closure()
            self._trainer.call_hook("on_before_clip_gradients")
            self._trainer.call_hook("clip_gadients")
            self._trainer.call_hook("on_after_clip_gadients")
        self.optimizer.step(closure=wrapper_closure)
```

Need to evaluate the limitations since clipping is currently tied to plugins

### Additional context

This would fix #5096, https://github.com/PyTorchLightning/pytorch-lightning/pull/6123#discussion_r587415267, https://github.com/PyTorchLightning/pytorch-lightning/issues/5671, https://github.com/PyTorchLightning/pytorch-lightning/issues/5982, and allow easily implementing new clipping techniques without having to merge them into Lightning

cc: @rohitgr7 who has been pushing for this for a while

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #6346

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
@carmocca should we deprecate `gradient_clip_val` and `gradient_clip_hook` arguments from Trainer? Also initial thoughts on this please :)

**Comment by ananthsub** [comment]
@rohitgr7 imo we should discuss this more on the issue. I think there are some design aspects that have to be worked out first

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 1]
```diff
@@ -478,7 +478,7 @@ def _training_step(self, split_batch: Any, batch_idx: int, opt_idx: int) -> Clos
 
         return result
 
-    def _track_and_norm_grad(self, optimizer: torch.optim.Optimizer) -> Dict[str, float]:
+    def _track_and_norm_grad(self, optimizer: torch.optim.Optimizer, opt_idx: int) -> Dict[str, float]:
```
Comment (rohitgr7) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:481]: off topic.. but there is a little inconsistency with argument name `opt_idx` here. In the hooks open to the users we have `optimizer_idx` but internally at some places, we have `opt_idx`.

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1473,6 +1478,43 @@ def untoggle_optimizer(self, optimizer_idx: int):
...
+    ):
+        """Perform Gradient Clipping for the optimizer parameters. Called before :meth:`optimizer_step`.
+
+        Args:
+            optimizer: Current optimizer being used. ``None`` if using manual optimization.
+            optimizer_idx: Index of the current optimizer being used. ``None`` if using manual optimization.
+            gradient_clip_val: The value at which to clip gradients. Passing ``gradient_clip_val=0`` disables gradient
+                clipping.
+            gradient_clip_algorithm: The gradient clipping algorithm to use. Pass ``gradient_clip_algorithm="value"``
+                for clip_by_value, and ``gradient_clip_algorithm="norm"`` for clip_by_norm.
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1496]: If it's a hook, the user doesn't call it, yet the docs here say that "passing =0 disables gradient clipping"? 
Meanwhile, this PR also deprecates the trainer arguments, so how is this feature supposed to be used now?

**Review by awaelchli — COMMENTED** [review]
I don't understand this PR. It seems to be incomplete. 
Today I am doing this: 

```
Trainer(gradient_clip_val=15.0)
```
or 
```
python train.py --gradient_clip_val 15.0
```

This PR aims at deprecating this option. How do I achieve the above now? There seems to be a hook attempting to replace this functionality, but this is missing the point why the Trainer argument exists in the first place. 99.9% of the time the user does not need to implement gradient clipping because Lightning can do it. But by removing this now, you force the user to  100% of the time have to implement a hook? IMO you are removing a perfectly fine feature and I vote against going in this direction. If configurability is desired, it should be entirely optional.

**Comment by rohitgr7** [comment]
so for now I have implemented option 2 from here: https://github.com/PyTorchLightning/pytorch-lightning/issues/6346#issuecomment-925914953

feel free to put your suggestions/recommendations here or on the issue :)

**Review by tchaton — APPROVED** [review]
LGTM ! Nice work :)

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1473,6 +1478,64 @@ def untoggle_optimizer(self, optimizer_idx: int):
         # save memory
         self._param_requires_grad_state = {}
 
+    def clip_gradients(self, optimizer: Optimizer, gradient_clip_val: Union[int, float], gradient_clip_algorithm: str):
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1481]: Can you scan the docs for a place where to advertise this hook?

Also add it to https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#hooks

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1473,6 +1478,64 @@ def untoggle_optimizer(self, optimizer_idx: int):
...
+        """Handles Gradient Clipping internally.
+
+        Args:
+            optimizer: Current optimizer being used.
+            gradient_clip_val: The value at which to clip gradients.
+            gradient_clip_algorithm: The gradient clipping algorithm to use. Pass ``gradient_clip_algorithm="value"``
+                for clip_by_value, and ``gradient_clip_algorithm="norm"`` for clip_by_norm.
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1488]:
```suggestion
                to clip by value, and ``gradient_clip_algorithm="norm"`` to clip by norm.
```

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1473,6 +1478,64 @@ def untoggle_optimizer(self, optimizer_idx: int):
...
+
+        Note:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1490]: Mind removing the "Note" and having it before `Args:`? It's an important detail

```python
        """Handles Gradient Clipping internally.

        <Move here>

        Args:
            ...
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1473,6 +1478,64 @@ def untoggle_optimizer(self, optimizer_idx: int):
...
+                will be available here.
+            gradient_clip_algorithm: The gradient clipping algorithm to use. By default value
+                passed in Trainer will be available here.
+
+        Example::
+
+            # Perform gradient clipping on discriminator (optimizer_idx=1) in GAN
+            def configure_gradient_clipping(self, optimizer, optimizer_idx, gradient_clip_val, gradient_clip_algorithm):
+                if optimizer_idx == 1:
+                    # lightning will handle the gradient clipping
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1528]:
```suggestion
                    # Lightning will handle the gradient clipping
```

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1473,6 +1478,64 @@ def untoggle_optimizer(self, optimizer_idx: int):
         # save memory
         self._param_requires_grad_state = {}
 
+    def clip_gradients(self, optimizer: Optimizer, gradient_clip_val: Union[int, float], gradient_clip_algorithm: str):
+        """Handles Gradient Clipping internally.
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1482]:
```suggestion
        """Handles gradient clipping internally.
```
the capitalization seems arbitrary here, or is it referred to as a name somewhere else too?

**Review by awaelchli — APPROVED** [review]
while this addition is probably needed and desired from a practical point of view, it also shows where Lightning's automatic optimization breaks down imo, because the desire of configuration here means ignoring settings that were provided through the trainer, which then in turn also challenges reproducibility.

**Comment by rohitgr7** [comment]
> while this addition is probably needed and desired from a practical point of view, it also shows where Lightning's automatic optimization breaks down imo, because the desire of configuration here means ignoring settings that were provided through the trainer, which then in turn also challenges reproducibility.

settings provided in the trainer will still work and one doesn't need to change their code at all since those trainer args are passed here by default. This is completely optional for those who want to configure it as per their requirement. Doesn't break automatic optimization.

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:1482]: na.. just looks better in docs.

**Comment by awaelchli** [comment]
Yes it does! Let me make it explicit with an example:

```
def configure_gradient_clipping(self,optimizer, optimizer_idx, gradient_clip_val, gradient_clip_algorithm)
    if optimizer_idx == 0:
        self.clip_gradients(optimizer, optimizer_idx, 10, "norm")
    else:
        self.clip_gradients(optimizer, optimizer_idx, 20, "norm")
```

Now:

```
trainer = Trainer(gradient_clip_val=30)  # has no effect!!!
```

Same issue when configuration is done externally:

```
python train.py --gradient_clip_val=30 
```

This is where Lightning reaches the limit and essentially breaks. This is an issue we will face more in the future.

**Comment by rohitgr7** [comment]
> n issue we will face more in th

okay.. so for this what we can do is make these two trainer arguments None by default. Now inside clip_gradients we can check:
1. if they are not None and different value is passed, then we can raise a warning/misconfigexception.
2. if they are not None and same value is passed, then it's fine.
3. if they are None and different value is passed, use the new value
4. if they are None, and no value is passed, then use the current default values that we have in Trainer (0.0 and 'norm')

wdyt @awaelchli ?

File: docs/source/common/lightning_module.rst [hunk 8]
```diff
@@ -1195,6 +1195,7 @@ for more information.
             on_after_backward()
 
             on_before_optimizer_step()
+            configure_gradient_clipping()
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:1198]: good that you added it here. because we discover here something not so nice. 

the `on_before_optimizer_step()` is now not running anymore directly before the optimizer step. do we want honor that and switch it up?

File: docs/source/common/optimizers.rst [hunk 9]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
     ):
         optimizer = optimizer.optimizer
         optimizer.step(closure=optimizer_closure)
+
+-----
+
+Configure Gradient Clipping
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:522]: the titles in this file are not capitalized

File: docs/source/common/optimizers.rst [hunk 10]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
+---------------------------
+To configure custom gradient clipping, consider overriding
+the :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_gradient_clipping` function.
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:525]: it's a method!

File: docs/source/common/optimizers.rst [hunk 11]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
+By default it will get arguments :attr:`~pytorch_lightning.trainer.Trainer.gradient_clip_val` and
+:attr:`~pytorch_lightning.trainer.Trainer.gradient_clip_algorithm` from ``Trainer`` in their respective
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:527]:
```suggestion
By default it will get arguments :attr:`~pytorch_lightning.trainer.trainer.Trainer.gradient_clip_val` and
:attr:`~pytorch_lightning.trainer.trainer.Trainer.gradient_clip_algorithm` from ``Trainer`` in their respective
```

File: docs/source/common/optimizers.rst [hunk 12]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
         optimizer = optimizer.optimizer
         optimizer.step(closure=optimizer_closure)
+
+-----
+
+Configure Gradient Clipping
+---------------------------
+To configure custom gradient clipping, consider overriding
+the :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_gradient_clipping` function.
+By default it will get arguments :attr:`~pytorch_lightning.trainer.Trainer.gradient_clip_val` and
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:526]: it will always get that, not just "by default".

File: docs/source/common/optimizers.rst [hunk 13]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
+:attr:`~pytorch_lightning.trainer.Trainer.gradient_clip_algorithm` from ``Trainer`` in their respective
+fields and lightning will handle gradient clipping on its own. In case you want to set different values
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:528]:
```suggestion
:attr:`~pytorch_lightning.trainer.Trainer.gradient_clip_algorithm` from ``Trainer`` in their respective
fields and Lightning will handle gradient clipping on its own. In case you want to set different values
```
fields?
You mean it will just get the values of the Trainer arguments as input. Can it be formulated without mentioning any "fields"?

File: docs/source/common/optimizers.rst [hunk 14]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
+for your arguments of your choice and let Lightning handle the gradient clipping, you can use the inbuilt
+:meth:`~pytorch_lightning.core.lightning.LightningModule.clip_gradients` function and pass the arguments
+along with your optimizer.
+
+.. note::
+    Make sure to not override :meth:`~pytorch_lightning.core.lightning.LightningModule.clip_gradients`
+    function. If you want to customize gradient clipping, consider using
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_gradient_clipping` function.
+
+For example, here we will apply gradient clipping only to optimizer A.
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:538]:
```suggestion
For example, here we will apply gradient clipping only to the gradients associated with optimizer A.
```

File: docs/source/common/optimizers.rst [hunk 15]
```diff
@@ -516,3 +516,49 @@ to perform a step, Lightning won't be able to support accelerators and precision
...
+            # Lightning will handle the gradient clipping
+            self.clip_gradients(
+                optimizer, gradient_clip_val=gradient_clip_val, gradient_clip_algorithm=gradient_clip_algorithm
+            )
+        elif optimizer_idx == 1:
+            custom_gradient_clip_val = 1e-2
+            for pg in optimizer.param_groups:
+                for p in pg["params"]:
+                    p.grad[p.grad > custom_gradient_clip_val] = custom_gradient_clip_val
+                    p.grad[p.grad <= 0] = 0
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:564]: that is a very funny way of clipping gradients haha
it will actually change the direction of the gradient
SGD will not be happy about this lol

Comment (awaelchli) [reply docs/source/common/optimizers.rst:564]: In all seriousness, I wouldn't put such an example.
maybe something like this could be more useful:

N = 2
# less clipping for parameters of optimizer 1
elif optimizer_idx == 1:
       self.clip_gradients(
                optimizer, gradient_clip_val=(N * gradient_clip_val), gradient_clip_algorithm=gradient_clip_algorithm
            )

or choose a different algorithm

File: docs/source/common/lightning_module.rst (hunk 8 above)
Comment (awaelchli) [reply docs/source/common/lightning_module.rst:1198]: EDIT: I see it may not be so easy to do it here. 
But still, I think it's slightly problematic

Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:1198]: yes.. i noticed.. so didn't change the sequence here.. maybe in another PR.
